### PR TITLE
pinecone_lookup

### PR DESCRIPTION
**Pull request by atbida**
This PR  
1) creates a folder for ML gems
2) adds a specific gem page for the new Pinecone Lookup Gem. NOTE: this format is a bit more lengthy compared with the OpenAI Gem doc page format. Feel free to suggest what aspects of each format are easier to read, etc.
3) adds some template pages for new gems (those pages have PRs which are coming separately)
4) moves the doc for conditional execution per Maciej's request

Updated the video to something acceptable.

### CODE REVIEW

**Comment by atbida** [comment]
To update with feedback noted here: https://coda.io/d/Prophecy-Product_dzpoYlcgcLl/ML-Gem-OpenAI_suS_B?searchClick=8bbc4722-a760-4c74-9fea-66d12138aabc_zpoYlcgcLl&search=#_lumOA

File: docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+tags: [generative-ai, machine-learning, llm, pinecone, openai]
+---
+
+<div class="wistia_responsive_padding" style={{padding:'56.25% 0 0 0', position:'relative'}}>
+<div class="wistia_responsive_wrapper" style={{height:'100%',left:0,position:'absolute',top:0,width:'100%'}}>
+<iframe src="https://fast.wistia.net/embed/iframe/019qt3csq9?seo=false?videoFoam=true" title="Getting Started With SQL Video" allow="autoplay; fullscreen" allowtransparency="true" frameborder="0" scrolling="no" class="wistia_embed" name="wistia_embed" msallowfullscreen width="100%" height="100%"></iframe>
+</div></div>
+<script src="https://fast.wistia.net/assets/external/E-v1.js" async></script>
+
+The Pinecone Lookup Gem identifies content that is similar to a provided vector embedding. The Gem operation, input, and output are breifly described below.
```
Comment (several27) [inline docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md:15]: Could we put more spacing here? 

Screenshot 2023-08-16 at 10 25 36 AM

File: docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+**Gem operation:** The Pinecone Lookup Gem calls the Pinecone API and returns a set of IDs with highest similarity to the provided embedding.
```
Comment (several27) [inline docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md:17]: What's the difference between the description and the "Gem Operation"? Not sure what is Gem operation?

Let's just have functionality description (no header needed) and then input and output description

File: docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+**Input:** This Gem requires an embedding as input. The embedding is provided by a foundational model like [OpenAI](https://platform.openai.com/docs/introduction).
```
Comment (several27) [inline docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md:19]: Maybe bullet points for better formatting?

File: docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+**Output:** This Gem outputs an array of IDs with corresponding similarity scores.
+
+![Input and Output](./img/pinecone_lookup_input_output.png)
+
+Now let’s understand the Gem Configuration, Input, and Output in detail.
+
+### Gem Configuration
```
Comment (several27) [inline docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md:27]: Everywhere else, we're calling this "Parameters" - let's stick to the same naming convention.

File: docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+| --------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------- |
+| model_embedding | array(float) - The format of this embedding is important. It must be an array of floating point numbers that matches the requirements of the Pinecone index. For example, we used a Pinecone index with `1536` dimensions, `Cosine` metric, and an `s1` pod type. So each record in the model_embedding column must be an array of `1536` floating point numbers between `-1 and 1`, such as `[-0.0018493991, -0.0059955865, ... -0.02498541]`. | True     |
```
Comment (several27) [inline docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md:53]: `floating point numbers between -1 and 1, such a` - is this accurate? 

This is definitely very common, but might not be accurate. You can have any ranged floats within the vectors (I believe).

File: docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        .withColumn("pinecone_matches", col("_response.matches"))\
+        .withColumn("pinecone_error", col("_response.error"))\
+        .drop("_vector", "_response")
+```
+
+</TabItem>
+<TabItem value="scala" label="Scala">
+
+```scala
+  [page under construction]
```
Comment (several27) [inline docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md:94]: You can just say Scala not supported.

File: docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+| model_embedding | array(float) - The format of this embedding is important. It must be an array of floating point numbers that matches the requirements of the Pinecone index. For example, we used a Pinecone index with `1536` dimensions, `Cosine` metric, and an `s1` pod type. So each record in the model_embedding column must be an array of `1536` floating point numbers between `-1 and 1`, such as `[-0.0018493991, -0.0059955865, ... -0.02498541]`. | True     |
+
+### Output
+
+The output Dataset contains the pinecone_matches column. For each input content entry, this Gem adds an array to the pinecone_matches column. The output array will have [Number of Results](/docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md#properties) entries.
```
Comment (several27) [inline docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md:57]: Contains matches and errors columns. First sentence is incorrect.

File: docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+![Configuration](./img/pinecone_lookup_configure.png)
+
+Verify the **(1) input columns** contain a column with the embeddings. The structure of this column's entries must be compatible with the structure of the Pinecone index.
+
+#### Credentials
+
+Configure the Pinecone API credentials here. Storing the Pinecone API token as a **(2) Databricks Secret** is highly recommended. For instructions click [here.](https://docs.databricks.com/en/security/secrets/index.html) Be sure to use the **(3) Fabric connection** to the Databricks workspace which contains the Databricks scope and secrets configured in this Gem.
+
+Hardcoding the Pinecone credential is not recommended. Selecting this option could send credentials to be stored hardcoded in Git; use only for credentials that should be shared with the world. Contact us to understand the integrations with other secret managers.
```
Comment (several27) [inline docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md:37]: Minor: They won't be shared with the world, but however, has git access.

**Review by several27 — CHANGES_REQUESTED** [review]
Requesting minor changes. Good stuff Anya

File: docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md (hunk 2 above)
Comment (atbida) [reply docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md:17]: Sounds good. I'll omit "Gem operation" from the ML gem docs. FYI here's the reason I had originally included the word "operation"
image

File: docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md (hunk 5 above)
Comment (atbida) [reply docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md:53]: You're right - both OpenAI and Pinecone DB allow a wider range of values!
https://platform.openai.com/docs/guides/embeddings/what-are-embeddings
https://www.pinecone.io/learn/vector-search-basics/

File: docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md (hunk 7 above)
Comment (atbida) [reply docs/low-code-spark/gems/machine-learning/ml-pinecone-lookup.md:57]: Updated to mention both pinecone_matches and pinecone_errors columns. Hopefully that's better?

**Review by several27 — APPROVED** [review]

**Comment by several27** [comment]
Looks great, thanks Anya!

**Pull request merged**